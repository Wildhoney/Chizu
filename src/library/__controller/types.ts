import { TypedError } from "../errors/utils.ts";
import { Models } from "../module/renderer/model/utils.ts";
import { Head } from "../module/renderer/types.ts";
import {
  Actions,
  Context,
  ContextTypes,
  Draft,
  Lifecycle,
  Meta,
  ModuleDefinition,
  Op,
} from "../types/index.ts";

export type ControllerActions<M extends ModuleDefinition> = {
  annotate<T>(value: T, operations?: (Op | Draft<T>)[]): T;
  produce(
    Æ’: (model: M["Model"], meta: Meta) => void,
  ): (models: Models<M["Model"]>, process: symbol) => Models<M["Model"]>;
  dispatch(action: M["Actions"]): Promise<void>;
  context<C extends Context>(context: C): ContextTypes<C>;
};

export type ControllerArgs<M extends ModuleDefinition> = Readonly<{
  model: Readonly<M["Model"]>;
  props: Readonly<M["Props"]>;
  actions: Readonly<ControllerActions<M>>;
}>;

export type ActionEvent<M extends ModuleDefinition> = (
  ...args: M["Actions"][number]
) => ActionGenerator<M>;

type ActionEvents<M extends ModuleDefinition> = {
  [K in Head<M["Actions"]>]: (
    payload: Payload<M["Actions"], K>,
  ) => ActionGenerator<M>;
};

export type ActionGenerator<M extends ModuleDefinition> =
  | void
  | Promise<void>
  | ((models: Models<M["Model"]>, process: symbol) => Models<M["Model"]>)
  | Promise<(models: Models<M["Model"]>, process: symbol) => Models<M["Model"]>>
  | AsyncGenerator<
      (models: Models<M["Model"]>, process: symbol) => Models<M["Model"]>,
      (models: Models<M["Model"]>, process: symbol) => Models<M["Model"]>,
      unknown
    >;

export type ControllerDefinition<M extends ModuleDefinition> = (
  controller: ControllerArgs<M>,
) => ControllerInstance<M>;

export type ControllerInstance<M extends ModuleDefinition> = {
  [Lifecycle.Mount]?(): ActionGenerator<M>;
  [Lifecycle.Derive]?(): ActionGenerator<M>;
  [Lifecycle.Node]?(tree: HTMLElement): ActionGenerator<M>;
  [Lifecycle.Error]?(error: Error | TypedError): ActionGenerator<M>;
  [Lifecycle.Unmount]?(): ActionGenerator<M>;
} & Partial<ActionEvents<M>>;

type Payload<A extends Actions, K> = A extends [K, infer P] ? P : never;
