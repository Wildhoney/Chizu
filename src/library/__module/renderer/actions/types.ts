import { UseBroadcast } from "../../../broadcast/types.ts";
import { UseContext } from "../../../context/types.ts";
import { ControllerArgs } from "../../../controller/types.ts";
import { ModuleDefinition } from "../../../types/index.ts";
import { ViewArgs } from "../../../view/types.ts";
import { UseOptions } from "../../types.ts";
import { UseDispatchers } from "../dispatchers/types.ts";
import { UseModel } from "../model/types.ts";
import { UseProps } from "../props/types.ts";

export type Props<M extends ModuleDefinition> = {
  broadcast: UseBroadcast;
  options: UseOptions<M>;
  model: UseModel;
  dispatchers: UseDispatchers;
  context: UseContext;
  props: UseProps;
};

export type UseActions<M extends ModuleDefinition> = {
  controller: ControllerArgs<M>;
  view: ViewArgs<M>;
};
