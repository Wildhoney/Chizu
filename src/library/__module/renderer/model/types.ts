import useModel from ".";
import { ModuleDefinition, State } from "../../../types/index.ts";
import { UseOptions } from "../../types.ts";

export type Props<M extends ModuleDefinition> = {
  options: UseOptions<M>;
};

export type UseModel = ReturnType<typeof useModel>;

export type Validate = {
  pending(): boolean;
  is(operation: State): boolean;
  draft<T>(): T;
};

export type Validatable<M> = {
  [K in keyof M]: M[K] extends object
    ? M[K] extends Array<infer U>
      ? Array<Validatable<U> & Validate> & Validate
      : Validatable<M[K]> & Validate
    : M[K] & Validate;
} & Validate;
