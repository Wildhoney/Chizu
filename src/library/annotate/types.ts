import { Annotation } from "./utils.ts";
import { Draft, Operation as Op, Process } from "../types/index.ts";

export type Operation<T> = {
  operation: Op | Draft<T>;
  process: Process;
};

export type Validateable<T> = T extends object
  ? {
      [P in keyof T]: T[P] extends (...args: unknown[]) => unknown
        ? T[P]
        : T[P] extends object
          ? Validateable<T[P]> | Annotation<T[P]>
          : Annotation<T[P]>;
    }
  : T;
