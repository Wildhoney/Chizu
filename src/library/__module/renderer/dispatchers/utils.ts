import { ActionEvent } from "../../../controller/types.ts";
import { intoError } from "../../../errors/utils.ts";
import { Lifecycle, ModuleDefinition, Task } from "../../../types/index.ts";
import { cleanup } from "../../../utils/produce/index.ts";
import { Head, Tail } from "../types.ts";
import { UseDispatchHandlerProps } from "./types.ts";

/**
 * @function useDispatcher
 * @param props {UseDispatchHandlerProps<M>}
 * @returns {(name: Head<M["Actions"]>, ƒ: GeneratorFn<M>) => (payload: Tail<M["Actions"]>) => Promise<void>}
 */
export function useDispatcher<M extends ModuleDefinition>(
  props: UseDispatchHandlerProps<M>,
) {
  return (_name: Head<M["Actions"]>, ƒ: ActionEvent<M>) => {
    return async (
      task: Task = Promise.withResolvers<void>(),
      payload: Tail<M["Actions"]>,
    ): Promise<void> => {
      if (typeof ƒ !== "function") return void task.resolve();

      // const abort = new AbortController();
      const process = Symbol("process");
      const action = ƒ(...payload);

      try {
        if (action == null) return void task.resolve();

        if (typeof action === "function") {
          const models = action(props.model.current, process);
          props.model.current = cleanup(models, process);
          props.update.rerender();

          return void task.resolve();
        }

        while (true) {
          if (action instanceof Promise) {
            const produce = await action;

            if (produce == null) return void task.resolve();

            const models = produce(props.model.current, process);
            props.model.current = cleanup(models, process);
            props.update.rerender();

            return void task.resolve();
          }

          const { value: produce, done } = await action.next();

          if (done) {
            const models = produce(props.model.current, process);
            props.model.current = cleanup(models, process);
            props.update.rerender();

            return void task.resolve();
          }

          props.model.current = produce(props.model.current, process);
          props.update.rerender();
        }
      } catch (error) {
        cleanup(props.model.current, process);
        props.update.rerender();
        props.unicast.emit(Lifecycle.Error, task, [intoError(error)]);
      }
    };
  };
}
