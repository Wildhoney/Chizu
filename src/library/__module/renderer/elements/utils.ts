import * as React from "react";

/**
 * @returns {string}
 */
export function useCacheKey(): string {
  const id = React.useId();
  const alphanumeric = id.replace(/[^a-zA-Z0-9]/g, "");

  return alphanumeric.replace(/\d/g, (digit) =>
    String.fromCharCode(103 + parseInt(digit, 10)),
  );
}
