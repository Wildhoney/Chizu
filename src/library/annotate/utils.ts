import traverse, { TraverseContext } from "traverse";
import { Model } from "../types";
import cloneDeep from "lodash/cloneDeep";
import { Store } from "../hooks/types.ts";
import * as React from "react";
import get from "lodash/get";
import { set } from "lodash";

import { Operations, Process } from "../types/index.ts";
import { Operation, Validator } from "./types.ts";

export class Annotation<T> {
  public operations: Operation<T>[];

  constructor(
    public value: T,
    operations: Operations<T>,
    process: Process,
  ) {
    this.operations = operations.map((operation) => ({
      operation,
      process,
    }));
  }

  merge(annotation: Annotation<T>): Annotation<T> {
    this.value = annotation.value;
    this.operations.push(...annotation.operations);
    return this;
  }

  update(value: T): Annotation<T> {
    this.value = value;
    return this;
  }

  clean(process: Process) {
    this.operations = this.operations.filter(
      (operation) => operation.process !== process,
    );
  }
}

export function reconcile<M extends Model>(
  model: M,
  produceModel: React.RefObject<M>,
  annotationStore: React.RefObject<Store>,
): M {
  const updatedProduceModel: M = <M>{};
  const updatedAnnotationStore: Store = {};

  const reconciled = traverse(cloneDeep(model)).forEach(function (
    this: TraverseContext,
  ): void {
    const path = this.path.join(".");
    const annotation = get(annotationStore.current, path);
    const key = this.path.join(".");

    if (!key) return;

    if (this.node instanceof Annotation) {
      if (annotation) {
        const merged = annotation.merge(this.node);
        updatedAnnotationStore[key] = merged;
        set(updatedProduceModel, key, merged.value);

        return void this.update(merged.value);
      } else {
        updatedAnnotationStore[key] = this.node;
        set(updatedProduceModel, key, this.node.value);

        return void this.update(this.node.value);
      }
    }

    if (annotation instanceof Annotation) {
      const updated = annotation.update(this.node);
      updatedAnnotationStore[key] = updated;
      set(updatedProduceModel, key, updated.value);

      return void this.update(updated.value);
    }

    updatedAnnotationStore[key] = this.node;
  });

  return (
    (annotationStore.current = updatedAnnotationStore),
    (produceModel.current = updatedProduceModel),
    reconciled
  );
}

export function proxy<M extends Model>(
  model: M,
  annotationStore: Store,
  path: string[] = [],
): M | Validator {
  return new Proxy(model, {
    get(target, property, receiver) {
      const key = [...path, property].join(".");
      const record = annotationStore?.[key];
      const pending = (record?.operations.length ?? 0) > 0;

      if (property === "pending") {
        return () => pending;
      }

      const value = Reflect.get(target, property, receiver);

      if (typeof value === "object" && value !== null)
        return proxy(value as M, annotationStore, [
          ...path,
          property as string,
        ]);
      else
        return {
          pending: () => pending,
          path: [...path, property],
        };
    },
  }) as M | Validator;
}
