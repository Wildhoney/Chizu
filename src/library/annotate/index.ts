import { Store } from "../hooks/types.ts";
import { Model, Operations, Process } from "../types/index.ts";
import { Annotation, proxy } from "./utils.ts";

export default function annotate<T>(
  value: T,
  operations: Operations<T>,
  process: Process,
) {
  return new Annotation<T>(value, operations, process) as unknown as T;
}

export function validateable<M extends Model>(
  model: M,
  annotationStore: Store,
) {
  return proxy(model, annotationStore);
}
