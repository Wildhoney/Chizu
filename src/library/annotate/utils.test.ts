import { describe, expect, it } from "@jest/globals";
import * as React from "react";
import { Store } from "../hooks/types";
import { reconcile } from "./utils";
import { State } from "../types";

import * as immer from "immer";
import { faker } from "@faker-js/faker";
import annotate from ".";

faker.seed(8_008);

const process = Symbol("process");

type Model = {
  country: string;
  countries: string[];
};

describe("reconcile()", () => {
  it("x", () => {
    const produceModel = React.createRef<Model>() as React.RefObject<Model>;
    const annotationStore = React.createRef<Store>() as React.RefObject<Store>;

    const a = {
      country: annotate(
        faker.location.country(),
        [State.Operation.Adding],
        process,
      ),
      countries: [
        faker.location.country(),
        annotate(faker.location.country(), [State.Operation.Adding], process),
        faker.location.country(),
      ],
    };

    const b = reconcile(a, produceModel, annotationStore);
    expect(b).toMatchSnapshot();
    expect(annotationStore.current).toMatchSnapshot();

    const c = immer.produce(b, (d) => {
      d.country = faker.location.country();
      d.countries[1] = annotate(
        faker.location.country(),
        [State.Operation.Updating],
        process,
      );
      d.countries.push(faker.location.country());
    });

    const d = reconcile(c, produceModel, annotationStore);
    expect(d).toMatchSnapshot();
    expect(annotationStore.current).toMatchSnapshot();
  });
});
