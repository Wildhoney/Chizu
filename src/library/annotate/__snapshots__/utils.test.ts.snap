// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reconcile() is able to handle the annotations 1`] = `
{
  "countries": [
    "Bahamas",
    "Nepal",
    "Samoa",
  ],
  "country": "Nigeria",
}
`;

exports[`reconcile() is able to handle the annotations 2`] = `
{
  "countries": [
    "Bahamas",
    "Nepal",
    "Samoa",
  ],
  "countries.0": "Bahamas",
  "countries.1": Annotation {
    "operations": [
      {
        "operation": 1,
        "process": Symbol(process),
      },
    ],
    "value": "Nepal",
  },
  "countries.2": "Samoa",
  "country": Annotation {
    "operations": [
      {
        "operation": 1,
        "process": Symbol(process),
      },
    ],
    "value": "Nigeria",
  },
}
`;

exports[`reconcile() is able to handle the annotations 3`] = `
{
  "countries": [
    "Bahamas",
    "Nepal",
    "Samoa",
    "Bonaire, Sint Eustatius and Saba",
  ],
  "country": "Venezuela",
}
`;

exports[`reconcile() is able to handle the annotations 4`] = `
{
  "countries": [
    "Bahamas",
    "Nepal",
    "Samoa",
    "Bonaire, Sint Eustatius and Saba",
  ],
  "countries.0": "Bahamas",
  "countries.1": Annotation {
    "operations": [
      {
        "operation": 4,
        "process": Symbol(process),
      },
      {
        "operation": 1,
        "process": Symbol(process),
      },
    ],
    "value": "Nepal",
  },
  "countries.2": "Samoa",
  "countries.3": "Bonaire, Sint Eustatius and Saba",
  "country": Annotation {
    "operations": [
      {
        "operation": 1,
        "process": Symbol(process),
      },
    ],
    "value": "Venezuela",
  },
}
`;
